// @nx/jest@17.3.2/src/generators/configuration/configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.configurationGenerator = void 0;
const init_1 = require("../init/init");
const check_for_test_target_1 = require("./lib/check-for-test-target");
const create_files_1 = require("./lib/create-files");
const create_jest_config_1 = require("./lib/create-jest-config");
const ensure_dependencies_1 = require("./lib/ensure-dependencies");
const update_tsconfig_1 = require("./lib/update-tsconfig");
const update_vscode_recommended_extensions_1 = require("./lib/update-vscode-recommended-extensions");
const update_workspace_1 = require("./lib/update-workspace");
const devkit_1 = require("@nx/devkit");
const js_1 = require("@nx/js");
const schemaDefaults = {
    setupFile: 'none',
    babelJest: false,
    supportTsx: false,
    skipSetupFile: false,
    skipSerializers: false,
    testEnvironment: 'jsdom',
};
function normalizeOptions(tree, options) {
    if (!options.testEnvironment) {
        options.testEnvironment = 'jsdom';
    }
    if (!options.hasOwnProperty('supportTsx')) {
        options.supportTsx = false;
    }
    // if we support TSX or compiler is not tsc, then we don't support angular(html templates)
    if (options.supportTsx ||
        options.babelJest ||
        ['swc', 'babel'].includes(options.compiler)) {
        options.skipSerializers = true;
    }
    if (options.skipSetupFile) {
        // setupFile is always 'none'
        options.setupFile = schemaDefaults.setupFile;
    }
    const project = (0, devkit_1.readProjectConfiguration)(tree, options.project);
    return {
        ...schemaDefaults,
        ...options,
        rootProject: project.root === '.' || project.root === '',
    };
}
async function configurationGenerator(tree, schema) {
    const options = normalizeOptions(tree, schema);
    const tasks = [];
    tasks.push(await (0, js_1.initGenerator)(tree, { ...schema, skipFormat: true }));
    tasks.push(await (0, init_1.jestInitGenerator)(tree, options));
    if (!schema.skipPackageJson) {
        tasks.push((0, ensure_dependencies_1.ensureDependencies)(tree, options));
    }
    await (0, create_jest_config_1.createJestConfig)(tree, options);
    (0, check_for_test_target_1.checkForTestTarget)(tree, options);
    (0, create_files_1.createFiles)(tree, options);
    (0, update_tsconfig_1.updateTsConfig)(tree, options);
    (0, update_vscode_recommended_extensions_1.updateVsCodeRecommendedExtensions)(tree);
    const nxJson = (0, devkit_1.readNxJson)(tree);
    const hasPlugin = nxJson.plugins?.some((p) => typeof p === 'string'
        ? p === '@nx/jest/plugin'
        : p.plugin === '@nx/jest/plugin');
    if (!hasPlugin) {
        (0, update_workspace_1.updateWorkspace)(tree, options);
    }
    if (!schema.skipFormat) {
        await (0, devkit_1.formatFiles)(tree);
    }
    return (0, devkit_1.runTasksInSerial)(...tasks);
}
exports.configurationGenerator = configurationGenerator;
exports.default = configurationGenerator;
